vowpalwabbit/core/src/reductions/cb/cb_to_cb_adf.cc

```diff
@@ -87,19 +87,46 @@ void predict_or_learn(cb_to_cb_adf& data, multi_learner& base, VW::example& ec)
   else { ec.pred.multiclass = data.adf_data.ecs[0]->pred.a_s[0].action + 1; }
 }
 
-void finish_example(VW::workspace& all, cb_to_cb_adf& c, VW::example& ec)
+void update_stats_cb_to_cb_adf(const VW::workspace& all, VW::shared_data& sd,
+    const cb_to_cb_adf& c, const VW::example& ec, VW::io::logger& logger)
 {
   if (c.explore_mode)
   {
-    c.adf_data.ecs[0]->pred.a_s = std::move(ec.pred.a_s);
-    c.adf_learner->print_example(all, c.adf_data.ecs);
+    c.adf_data.ecs[0]->pred.a_s = ec.pred.a_s;
   }
   else
   {
-    c.adf_data.ecs[0]->pred.multiclass = std::move(ec.pred.multiclass);
-    c.adf_learner->print_example(all, c.adf_data.ecs);
+    c.adf_data.ecs[0]->pred.multiclass = ec.pred.multiclass;
   }
-  VW::finish_example(all, ec);
+  c.adf_learner->update_stats(all, sd, &c, c.adf_data.ecs, logger);
```

Comment:
`adf_learner`'s data is not the c, but its own data object. Leaving data as private in the learner interface will fix this issue